OpenRA.Mods.Common/Widgets/Selection.cs

```diff
@@ -0,0 +1,245 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Primitives;
+
+namespace OpenRA.Mods.Common.Widgets
+{
+	class Selection
+	{
+		public enum States { Empty, Ready, Active, Inactive }
+		public States State = States.Empty;
+		public LabelWithSelectionWidget SelectionWidget = null;
+		public int2 StartingLocation;
+		public int2 EndingLocation;
+
+		RenderedText TextAsDrawn
+		{
+			get
+			{
+				return RenderedText.From(
+						SelectionWidget.GetTextContent(),
+						SelectionWidget.GetFont(),
+						SelectionWidget.GetPosition());
+			}
+		}
+
+		public int Start
+		{
+			get
+			{
+				return Math.Min(
+						TextAsDrawn.ClosestCharacter(StartingLocation).Index,
+						TextAsDrawn.ClosestCharacter(EndingLocation).Index);
+			}
+		}
+
+		public int End
+		{
+			get
+			{
+				return Math.Max(
+						TextAsDrawn.ClosestCharacter(StartingLocation).Index,
+						TextAsDrawn.ClosestCharacter(EndingLocation).Index);
+			}
+		}
+
+		public Selection() { }
+
+		public bool OwnedBy(LabelWidget widget)
+		{
+			return SelectionWidget == widget;
+		}
+
+		public bool HandleMouseDown(LabelWithSelectionWidget widget, int2 location)
+		{
+			if (object.ReferenceEquals(widget, null))
+				return false;
+
+			switch (State)
+			{
+				case States.Empty:
+				case States.Inactive:
+					SelectionWidget = widget;
+					StartingLocation = location;
+					EndingLocation = location;
+					State = States.Ready;
+					return true;
+				default:
+					return false;
+			}
+		}
+
+		public bool HandleMouseMove(int2 location)
+		{
+			if (object.ReferenceEquals(SelectionWidget, null))
+				return false;
+
+			switch (State)
+			{
+				case States.Ready:
+					EndingLocation = location;
+					State = States.Active;
+					return true;
+				case States.Active:
+					EndingLocation = location;
+					return true;
+				default:
+					return false;
+			}
+		}
+
+		public bool HandleMouseUp()
+		{
+			switch (State)
+			{
+				case States.Active:
+					State = States.Inactive;
+					return true;
+				case States.Ready:
+					State = States.Empty;
+					SelectionWidget = null;
+					return true;
+				default:
+					return false;
+			}
+		}
+
+		public bool HandleLooseMouseFocus()
+		{
+			State = States.Empty;
+			SelectionWidget = null;
+			return true;
+		}
+
+		class RenderedText
+		{
+			public class Character
+			{
+				// NOTE: What happens when we handle unicode?
```

Comment:
We will need to answer this before we can merge: I suspect the answer will be "break". I don't yet have a feeling for how this is used, but in general assuming that "text character" == `char` is a bad move when working with modern text.